Revit_Core_Engine/Compute/CachePanelGeometry.cs

```diff
@@ -0,0 +1,84 @@
+using BH.oM.Adapters.Revit.Enums;
+using BH.oM.Adapters.Revit.Settings;
+using BH.oM.Base;
+using BH.oM.Base.Attributes;
+using BH.oM.Geometry;
+using BH.Revit.Engine.Core.Objects;
+using System.Collections.Generic;
+using System.ComponentModel;
+
+namespace BH.Revit.Engine.Core
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Extracts the panel geometry from Revit walls, floors, roofs etc. and caches it in refObjects in order to use it in panel converts downstream." +
+                     "\nIt is an optimisation meant to minimise the number of document regenerations on panel converts, which often become the major time consumer.")]
+        [Input("document", "Revit document hosting the panel elements with locations to be cached.")]
+        [Input("elementIds", "Ids of the panel elements with locations to be cached.")]
+        [Input("discipline", "Engineering discipline based on the BHoM discipline classification.")]
+        [Input("settings", "Revit adapter settings to be used while performing the query.")]
+        [Input("refObjects", "Collection of objects already processed in the current adapter action, stored to avoid processing the same object more than once." +
+               "\nExtracted location surfaces are stored here.")]
+        public static void CachePanelGeometry(this Document document, List<ElementId> elementIds, Discipline discipline, RevitSettings settings, Dictionary<string, List<IBHoMObject>> refObjects)
+        {
+            List<HostObject> hostObjectsToExtractSurfaces = new List<HostObject>();
+            foreach (ElementId id in elementIds)
+            {
+                HostObject hostObject = document.GetElement(id) as HostObject;
+                if (hostObject == null)
+                    continue;
+
+                bool outlinesFound = false;
+                if (discipline == Discipline.Structural)
+                {
+                    List<ICurve> outlines = hostObject.AnalyticalOutlines(settings);
+                    if (outlines != null && outlines.Count != 0)
+                    {
+                        outlinesFound = true;
+                        OutlineCache cache = new OutlineCache { Outlines = outlines };
+                        refObjects.Add(id.OutlineCacheKey(), new List<IBHoMObject> { cache });
+                    }
+                }
+
+                if (!outlinesFound)
+                    hostObjectsToExtractSurfaces.Add(hostObject);
+            }
+
+            var surfacesToCache = hostObjectsToExtractSurfaces.PanelSurfaces(discipline, settings);
+            foreach (HostObject hostObject in hostObjectsToExtractSurfaces)
+            {
+                SurfaceCache cache = new SurfaceCache { Surfaces = surfacesToCache[hostObject.Id] };
```

Comment:
Somehow I got an exception here 😅 
![image](https://user-images.githubusercontent.com/102604891/185153517-26ba21d3-0708-4009-83dd-5f694b89885d.png)

![image](https://user-images.githubusercontent.com/102604891/185153589-688883bc-8307-43ea-a149-cd077714581e.png)

Reply:
Interestingly enough, it worked on my local yesterday - must have broken this when splitting the code into commits and adding final tweaks. Interesting bug related to invalidation of Revit `Element` objects in time. Should be good now 👍

Reply:
It works! However, would it be simpler to do `IsValidObject`?
```c#
foreach (HostObject hostObject in hostObjectsToExtractSurfaces)
{
    if (hostObject.IsValidObject)
        continue;

    SurfaceCache cache = new SurfaceCache { Surfaces = surfacesToCache[hostObject.Id] };
    refObjects.Add(hostObject.Id.SurfaceCacheKey(), new List<IBHoMObject> { cache });
}
```

Reply:
No, because the surfaces were extracted correctly and there is nothing wrong with the Revit element itself - it is only the object in the memory being invalidated due to the temp transaction that has been rolled back. So skipping these surfaces would simply mean them being extracted again on the actual convert, for no practical reason.